UM/Qt/QtRenderer.py

```diff
@@ -0,0 +1,192 @@
+        return round(Application.getInstance().primaryScreen().physicalDotsPerInch() / 96.0)
+
+    ##  Get the list of render batches.
+    def getBatches(self):
+        return self._batches
+
+    ##  Overridden from Renderer.
+    #
+    #   This makes sure the added render pass has the right size.
+    def addRenderPass(self, render_pass):
+        super().addRenderPass(render_pass)
+        render_pass.setSize(self._viewport_width, self._viewport_height)
+
+    ##  Set background color used when rendering.
+    def setBackgroundColor(self, color):
+        self._background_color = color
+
+    ##  Set the viewport size.
+    #
+    #   \param width The new width of the viewport.
+    #   \param height The new height of the viewport.
+    def setViewportSize(self, width, height):
+        self._viewport_width = width
+        self._viewport_height = height
+
+        for render_pass in self._render_passes:
+            render_pass.setSize(width, height)
+
+    ##  Set the window size.
+    def setWindowSize(self, width, height):
+        self._window_width = width
+        self._window_height = height
+
+    ##  Get the window size.
+    #
+    #   \return A tuple of (window_width, window_height)
+    def getWindowSize(self):
+        return (self._window_width, self._window_height)
+
+    def beginRendering(self):
+        if not self._initialized:
+            self._initialize()
+
+        self._gl.glViewport(0, 0, self._viewport_width, self._viewport_height)
+        self._gl.glClearColor(self._background_color.redF(), self._background_color.greenF(), self._background_color.blueF(), self._background_color.alphaF())
+        self._gl.glClear(self._gl.GL_COLOR_BUFFER_BIT | self._gl.GL_DEPTH_BUFFER_BIT)
+        self._gl.glClearColor(0.0, 0.0, 0.0, 0.0)
+
+    ##  Put a node in the render queue
+    def queueNode(self, node, **kwargs):
+        type = RenderBatch.RenderType.Solid
+        if kwargs.get("transparent", False):
+            type = RenderBatch.RenderType.Transparent
+        elif kwargs.get("overlay", False):
+            type = RenderBatch.RenderType.Overlay
+
+        batch = RenderBatch(
+            kwargs.get("shader", self._default_material),
+            type = type,
+            mode = kwargs.get("mode", RenderBatch.RenderMode.Triangles),
+            backface_cull = kwargs.get("backface_cull", True),
+            range = kwargs.get("range", None)
+        )
+
+        batch.addItem(node.getWorldTransformation(), kwargs.get("mesh", node.getMeshData()))
+
+        self._batches.append(batch)
+    
+    ##  Render all nodes in the queue
+    def renderQueuedNodes(self):
+        self._batches.sort()
+
+        for render_pass in self.getRenderPasses():
+            render_pass.render()
+
+    def endRendering(self):
+        self._batches.clear()
+
+    ##  Render a full screen quad.
+    #
+    #   \param shader The shader to use when rendering.
+    def renderFullScreenQuad(self, shader):
```

Comment:
OpenGL by definition can only render with one active shader program. If you need to render quads with multiple shaders you will need to call renderFullScreenQuad multiple times.